### PR TITLE
Efficiently finish sorting by (x, y) stream, that already sorted by (x) [CLICKHOUSE-4022]

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
TODO:
- style errors;
- rename `FinishMergeSortingBlockInputStream` to `FinishSortingBlockInputStream` because it doesn't merge anything.

File: dbms/src/DataStreams/MergeSortingBlockInputStream.cpp [hunk 1]
```diff
@@ -304,4 +305,138 @@ void MergeSortingBlockInputStream::remerge()
...
+            }
+
+            /// If we reach here, that means that current block is first in chunk
+            /// or it all consists of rows with the same key as tail of a previous block.
+            blocks.push_back(block);
+        }
+
+        /// We need to save tail of block, because next block may starts with the same key as in tail
+        /// and we should sort these rows in one chunk.
+        if (block)
```
Comment (ztlpn) [inline dbms/src/DataStreams/MergeSortingBlockInputStream.cpp]: Suggestion: maybe it is better to move this code inside the loop before `break` - this way it is more evident that the block is cut in two only if it contains the next key.

File: dbms/src/DataStreams/MergeSortingBlockInputStream.h [hunk 2]
```diff
@@ -131,4 +131,45 @@ class MergeSortingBlockInputStream : public IProfilingBlockInputStream
     bool remerge_is_useful = true;
 };
 
+
+/** Takes stream already sorted by `x` and finishes sorting it by (`x`, `y`).
```
Comment (ztlpn) [inline dbms/src/DataStreams/MergeSortingBlockInputStream.h]: Looks like the stream expects that each block in the stream is individually sorted by `description_to_sort` but it isn't reflected in the comment.

File: dbms/src/DataStreams/MergeSortingBlockInputStream.cpp [hunk 3]
```diff
@@ -304,4 +305,138 @@ void MergeSortingBlockInputStream::remerge()
...
+    const BlockInputStreamPtr & input, SortDescription & description_sorted_,
+    SortDescription & description_to_sort_,
+    size_t max_merged_block_size_, size_t limit_)
+    : description_sorted(description_sorted_), description_to_sort(description_to_sort_),
+    max_merged_block_size(max_merged_block_size_), limit(limit_)
+{
+    children.push_back(input);
+    header = children.at(0)->getHeader();
+    removeConstantsFromSortDescription(header, description_sorted);
+    removeConstantsFromSortDescription(header, description_to_sort);
```
Comment (ztlpn) [inline dbms/src/DataStreams/MergeSortingBlockInputStream.cpp]: We should check that `description_sorted` is a prefix of `description_to_sort`

File: dbms/src/DataStreams/MergeSortingBlockInputStream.h (hunk 2 above)
Comment (ztlpn) [reply dbms/src/DataStreams/MergeSortingBlockInputStream.h]: BTW there is no mention of that in the task. Or is this stream intended to use always in conjunction with `PartialSortingBlockInputStream(description_to_sort)`?

**Review by ztlpn — CHANGES_REQUESTED** [review]

File: dbms/src/DataStreams/tests/finish_sorting_stream.cpp [hunk 4]
```diff
@@ -96,9 +97,9 @@ int main(int argc, char ** argv)
                 << std::endl;
         }
     }
-    catch (const Exception & e)
```
Comment (ztlpn) [inline dbms/src/DataStreams/tests/finish_sorting_stream.cpp]: Doesn't compile without these lines :)

File: dbms/src/DataStreams/tests/finish_sorting_stream.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        for (auto & block : blocks)
+            sortBlock(block, sort_descr);
+
+        BlockInputStreamPtr stream = std::make_shared<MergeSortingBlocksBlockInputStream>(blocks, sort_descr, n);
+
+        SortDescription sort_descr_final;
+        sort_descr_final.emplace_back("col1", 1, 1);
+        sort_descr_final.emplace_back("col2", 1, 1);
+
+        stream = std::make_shared<PartialSortingBlockInputStream>(stream, sort_descr_final);
```
Comment (ztlpn) [inline dbms/src/DataStreams/tests/finish_sorting_stream.cpp]: PartialSorting is unnecessary now.

File: dbms/src/DataStreams/tests/finish_sorting_stream.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                }
+
+            stopwatch.stop();
+            std::cout << std::fixed << std::setprecision(2)
+                << "Elapsed " << stopwatch.elapsedSeconds() << " sec."
+                << ", " << n / stopwatch.elapsedSeconds() << " rows/sec."
+                << std::endl;
+        }
+    }
+    {
```
Comment (ztlpn) [inline dbms/src/DataStreams/tests/finish_sorting_stream.cpp]: Catch clause is still necessary :)

File: dbms/src/DataStreams/FinishSortingBlockInputStream.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+FinishSortingBlockInputStream::FinishSortingBlockInputStream(
+    const BlockInputStreamPtr & input, const SortDescription & description_sorted_,
+    const SortDescription & description_to_sort_,
+    size_t max_merged_block_size_, size_t limit_)
+    : description_sorted(description_sorted_), description_to_sort(description_to_sort_),
+    max_merged_block_size(max_merged_block_size_), limit(limit_)
+    
+{
+    if (!isPrefix(description_sorted, description_to_sort))
+        throw Exception("Can`t finish sorting. SortDescription of already sorted streamis not prefix of SortDescription needed to sort", ErrorCodes::LOGICAL_ERROR);
```
Comment (ztlpn) [inline dbms/src/DataStreams/FinishSortingBlockInputStream.cpp]:
```suggestion
        throw Exception("Can`t finish sorting. SortDescription of already sorted stream is not prefix of"
            "SortDescription needed to sort", ErrorCodes::LOGICAL_ERROR);
```

File: dbms/src/DataStreams/FinishSortingBlockInputStream.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        if (end_of_stream)
+            return {};
+
+        blocks.clear();
+        if (tail_block)
+            blocks.push_back(std::move(tail_block));
+
+        Block block;
+        size_t tail_pos = 0;
```
Comment (ztlpn) [inline dbms/src/DataStreams/FinishSortingBlockInputStream.cpp]: `block` and `tail_pos` can be moved closer to where they are used

**Review by ztlpn — APPROVED** [review]
Everything looks ok, a couple of small corrections are still needed.

File: dbms/src/DataStreams/FinishSortingBlockInputStream.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            else if (res > 0)
+                return false;
+        }
+        return false;
+    }
+};
+
+Block FinishSortingBlockInputStream::readImpl()
+{
+    if (limit && total_rows_processed == limit)
```
Comment (alexey-milovidov) [inline dbms/src/DataStreams/FinishSortingBlockInputStream.cpp]: Maybe this is correct, but I prefer `>=` to be sure.

**Comment by nvartolomei** [comment]
Keeping in mind how columns are resolved during query parsing, the following query might change it's behaviour if FinishSortingBlockInputStream is plumbed as is.

```sql
select -a as a, b
from (
      select a, b
      from (
            select number as a, rand() as b from system.numbers limit 10
             )
      order by a)
order by a, b;
```

SortDescription equality in that case will be true, right?

**Comment by ztlpn** [comment]
FinishSortingBlockInputStream is intended as a building block for implementing ORDER BY clause execution that uses the ordering in MergeTree parts. Your example (at least for now) will just be sorted twice.

**Comment by nvartolomei** [comment]
@ztlpn I was hoping this is a building block for #1344 :)

**Pull request merged**